### PR TITLE
MGMT-16263: Enable usage of custom OCP release images when we create cluster

### PR DESCRIPTION
**Pull request by danmanor**
## In General
This PR is part 2 of the Implementation of this enhancement proposal - https://github.com/openshift/assisted-service/blob/master/docs/enhancements/dynamic-fetching-of-openshift-versions.md, which is enabling usage of dynamic `OpenShift` releases. (part 1 - https://github.com/openshift/assisted-service/pull/5916)

### Main Flow
As mentioned in the description in part 1, the main changes are:

1. **In openshift-versions endpoint** - We retrieve the release images from the DB (from part (1), which contains static and dynamic release images), and in addition we filter by 2 new query parameters (only_latest, version_pattern) that will be used by the `UI` to present short list of the versions. We also filter by a new configuration parameter `IGNORED_OPENSHIFT_VERSIONS`, which contains a list of versions we don't want to retrieve (as one might have a blocker bug).

2. **We split the versions handler** to 2 different version handlers - one for `RestAPI` and one for `KubeAPI`. Each of the structs will implement version handling according to the running mode.

### Remarks

1. I decided to split the version handler into two separate implementations for `RestAPI` and `KubeAPI` because the changes necessitated distinct handling of versions in each mode.
2. This change will trigger release images syncer to start in `SaaS` .
3. The current process for subsystem testing begins with deploying `Wiremock` (to mock `OCM`), then launching the assisted service and other components, and finally starting the tests. Stubs are added to `Wiremock` after the test initiation, which poses a problem. This is because part 1 (see [part 1](https://github.com/openshift/assisted-service/pull/5916)) implements a utility that needs to interact with `OCM` immediately upon the service's startup. This necessitates having the stubs available in `Wiremock` from the moment the service begins (to retrieve releases). To solve this, I have implemented a Python script that adds the stubs to `Wiremock`, ensuring they are in place before the service starts.

## List all the issues related to this PR

- [x] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [x] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: Dockerfile.assisted-service
Comment (danielerez) [inline Dockerfile.assisted-service:40]: It's a bit non-conservative to include a binary for tests purposes in the production image. Not sure if we can easily avoid it though, but maybe just worth thinking about alternatives.

File: data/default_release_sources.json
Comment (danielerez) [inline data/default_release_sources.json:1]: I haven't followed the original change, but I just wonder why is this object separated from default_release_images.json? Is it just for simplicity, or is there some other consideration? Also, do we handle updating these values with the upgrade script (i.e. when a new OCP version is released, etc)?

File: internal/versions/rest_api_versions.go
Comment (danielerez) [inline internal/versions/rest_api_versions.go:84]: Instead of escaping the version and use sql, won't it be simpler to find relevant release images with something like gofunk? E.g. https://github.com/openshift/assisted-service/blob/cc56ded8ef0a9dc3f95e2c8a65d08e96546cd72b/internal/versions/versions.go#L175

**Review by danielerez — COMMENTED** [review]
Partial review, add some comments for now.

File: Dockerfile.assisted-service
Comment (danmanor) [reply Dockerfile.assisted-service:40]: I don't think it would be complicated, but I don't think it worth the maintaining of another image

File: cmd/main.go
Comment (danielerez) [inline cmd/main.go:280]: Consider logging also the IgnoredOpenshiftVersions, could be useful for debugging,

File: internal/bminventory/inventory.go
Comment (danielerez) [inline internal/bminventory/inventory.go:469]: We're no longer handling 'maintenance' support level?

**Review by danielerez — COMMENTED** [review]
Looks good overall, added some comments.
Just a general comment, as discussed, please check disconnected scenarios (i.e. disable sync to db, etc).

File: internal/bminventory/inventory.go
Comment (danmanor) [reply internal/bminventory/inventory.go:469]: We do, but for some reason there was a limitation in the service code that prohibited using maintenance release images for installing cluster, I removed it because we want to support this option.

File: internal/versions/rest_api_versions.go
Comment (danmanor) [reply internal/versions/rest_api_versions.go:179]: In DB actions it is clearer for me to check the error in a different line to distinguish the action

File: internal/versions/rest_api_versions.go
Comment (danmanor) [reply internal/versions/rest_api_versions.go:131]: I was thinking about that, but I think it I can't do much validation because urls can be in many forms (in the CI it is much different for example). I think if it will be empty it just won't find anything, do you think it worth checking it separately ?

**Comment by danmanor** [comment]
@danielerez @gamli75 holding until the UI will merge the adjustments to this change

**Review by carbonin — CHANGES_REQUESTED** [review]
Holding this to take a look at the wiremock situation.

I very much don't like the idea of including non-production code that can alter/intercept network communication in our production image.

**Comment by carbonin** [comment]
Also just as a general note, splitting into separate commits is great, but try to make the commit messages usable.
All this prefixing is unnecessary:
image

File: Dockerfile.assisted-service
Comment (carbonin) [reply Dockerfile.assisted-service:40]: We're already maintaining the image we're deploying for wiremock (or should be at least).
[This](https://github.com/openshift/assisted-service/blob/3d14207ff4155d3f982d06fd485c828b0f0594f4/deploy/wiremock/wiremock-deployment.yaml#L18) shows it's using `ocpmetal` repo which IIRC is something we used to maintain.

Comment (carbonin) [reply Dockerfile.assisted-service:40]: FWIW I tried to get rid of all of these in b8aa0b95615847c42434c46b6a08678b3d5d4476 but I guess I missed some

Comment (carbonin) [reply Dockerfile.assisted-service:40]: Also I'm not sure why converting to python is the only option.

In the past I was able to initialize the wiremock instance using a small go program. Could we not do the same as a part of deploying wiremock (before we start assisted-service)?

Example:
```
package main

import (
	"log"

	"github.com/kelseyhightower/envconfig"
	"github.com/openshift/assisted-service/subsystem"
)

var Options struct {
	OCMHost   string `envconfig:"OCM_HOST"`
	TestToken string `envconfig:"TEST_TOKEN"`
}

func main() {
	if err := envconfig.Process("subsystem", &Options); err != nil {
		log.Fatal(err.Error())
	}

	wiremock := &subsystem.WireMock{
		OCMHost:   Options.OCMHost,
		TestToken: Options.TestToken,
	}

	if err := wiremock.DeleteAllWiremockStubs(); err != nil {
		log.Fatalf("Failed to delete all wiremock stubs, %v", err)
	}

	if err := wiremock.CreateWiremockStubsForOCM(); err != nil {
		log.Fatalf("Failed to init wiremock stubs, %v", err)
	}
}
```

Comment (danmanor) [reply Dockerfile.assisted-service:40]: @carbonin IIUC this is what I did exactly

**Comment by danmanor** [comment]
@carbonin This code runs only when we run subsystem test though, then why can it interfere with the production flow ?

File: internal/versions/rest_api_versions_test.go
Comment (carbonin) [inline internal/versions/rest_api_versions_test.go:1]: Can you ensure you retain the tests added in https://github.com/openshift/assisted-service/pull/6066/files#diff-83dbd9742dbbf03982097164da597f3c30dcda4ed0af066409af7f3ac7f67716R281-R330 to ensure we don't regress?

File: internal/versions/test_common.go
Comment (carbonin) [inline internal/versions/test_common.go:1]: Should this be `common_test.go`? All test files should be `*_test.go`, right?

File: subsystem/agent_based_installer_client_test.go
Comment (carbonin) [inline subsystem/agent_based_installer_client_test.go:29]: Why was this changed?

File: subsystem/operators_test.go
Comment (carbonin) [inline subsystem/operators_test.go:249]: Why change this?

File: swagger.yaml
Comment (carbonin) [inline swagger.yaml:3119]: You would literally put `4.12` in here? Is that how that works?

Would it be better for this to be an enum with `X.Y` and `X.Y.Z` as members or are there more options?

File: tools/deploy_assisted_installer_configmap.py
Comment (carbonin) [inline tools/deploy_assisted_installer_configmap.py:115]: Why change this? Every other field is the previous way.

**Review by carbonin — COMMENTED** [review]
I would also like us to try to look more closely for solutions to tests that don't involve wiremock stuff in our prod image.

File: Dockerfile.assisted-service
Comment (carbonin) [reply Dockerfile.assisted-service:40]: Why not do it from outside the assisted service image?

File: internal/versions/kube_api_versions_test.go
Comment (danmanor) [reply internal/versions/kube_api_versions_test.go:221]: I didn't understand what was this test trying to test, but I get it now. I will insert it back

File: internal/versions/kube_api_versions_test.go
Comment (danmanor) [reply internal/versions/kube_api_versions_test.go:431]: I think we can be verbose in this one and keep it in separate tests as it uses `GetReleaseImage` for specific instance

Comment (carbonin) [reply internal/versions/kube_api_versions_test.go:431]: But the content is exactly the same.
I don't see value in having to change identical tests in two places when we make one code change unless there is significant differences between how the function is used, which is not the case here.

File: internal/versions/rest_api_versions.go
Comment (danmanor) [reply internal/versions/rest_api_versions.go:58]: The thought process was as follows:
For KubeAPI mode, we want RELEASE_IMAGES to serve as cache.
For RestAPI mode, there are two scenarios:

1. In SaaS, we want to activate the release images gathering process, and use the DB as the source for release images.
2. In other scenarios (ABI for example), we don't want to activate the release images gathering, (by not supplying RELEASE_SOURCES, then we want to use RELEASE_IMAGES as the source for release images.

This is the way I implemented it. An alternative could be having the release images in the DB in rest mode anyway, and if the gathering is activated then add more releases, otherwise not. then you can only look at the DB for the source of release images in Rest mode, although I found this solution much bigger change then the current option.

File: internal/versions/rest_api_versions_test.go
Comment (danmanor) [reply internal/versions/rest_api_versions_test.go:1]: Yes I added in kubeAPI mode, in rest we don't use that code

File: subsystem/operators_test.go
Comment (danmanor) [reply subsystem/operators_test.go:249]: It needs to match the version inside `../docs/hive-integration/crds/clusterImageSet.yaml`, which I upgraded

File: swagger.yaml
Comment (danmanor) [reply swagger.yaml:3119]: Currently it retrieves every release image such that `version_pattern` is substring of `ReleaseImage.Version`

File: subsystem/agent_based_installer_client_test.go
Comment (danmanor) [reply subsystem/agent_based_installer_client_test.go:29]: It needs to match the version inside `../docs/hive-integration/crds/clusterImageSet.yaml`, which I upgraded

File: subsystem/operators_test.go
Comment (danmanor) [reply subsystem/operators_test.go:249]: We register a cluster with `s390x` arch, we should get this arch now

File: internal/versions/rest_api_versions.go
Comment (danmanor) [reply internal/versions/rest_api_versions.go:58]: In second thought, it might be better to go with the alternative and have one source of release images for rest API, I will try it out

File: internal/versions/kube_api_versions_test.go
Comment (danmanor) [reply internal/versions/kube_api_versions_test.go:431]: I tried to convert it, in my opinion it makes less sense in common as the test itself doesn't have much context but calling the function with different parameters, therefore it turns up as a bunch of restAPI calls (with restAPI.GetReleaseImage) and then a bunch of kubeAPI calls, same as if they were in separate files. In my opinion, it would make more sense to keep them as they are currently, but if you disagree I can merge the tests in common_test.go

Comment (carbonin) [reply internal/versions/kube_api_versions_test.go:397]: From meeting: You can "mock" the `GetReleaseImage` by passing in a function defined just for the test to return some test release image. Don't worry about the different real implementations of `GetReleaseImage`

File: swagger.yaml
Comment (carbonin) [reply swagger.yaml:3119]: Okay if we want this then we should change the name.
I'd call it just `version` assuming it will return all the data for the versions matching whatever is sent.

File: subsystem/operators_test.go
Comment (carbonin) [reply subsystem/operators_test.go:249]: I'm not sure if this was being done intentionally. If the test wasn't failing before I'm worried it was behavior we were somehow expecting.

File: Dockerfile.assisted-service
Comment (carbonin) [reply Dockerfile.assisted-service:40]: Lets try to just run the go code to initialize the mocks from the Makefile.
If that doesn't work we can explore other options.

File: cmd/main.go
Comment (carbonin) [inline cmd/main.go:446]: Is this a problem that this runs on every pod (not just the leader)?
I assume we'll get duplicates or errors when we try to insert the same version from multiple replicas.

Maybe do this in the release syncer process?

File: cmd/main.go
Comment (carbonin) [inline cmd/main.go:987]: I think we also need to check release sources here.
Because we want to just let the syncer thread add these if it's going to run rather than adding them here and then having the sync thread immediately delete and readd them.

File: deploy/wiremock/wiremock-deployment.yaml
Comment (carbonin) [inline deploy/wiremock/wiremock-deployment.yaml:18]: Just for reference, we should figure out how this was built and if we should update it or move it to `edge-infrastructure` I don't think we do anything with `ocpmetal` anymore.

cc @gamli75 what happened with that org? Do we even own it still?

File: docs/hive-integration/crds/clusterImageSet.yaml
Comment (carbonin) [inline docs/hive-integration/crds/clusterImageSet.yaml:4]: Why update this now? Everything kubeapi related should remain the same, right?

File: internal/releasesources/clients.go
Comment (carbonin) [inline internal/releasesources/clients.go:36]: If we are able to remove the wiremock stuff I don't think this should be exported.

File: internal/versions/api.go
Comment (carbonin) [inline internal/versions/api.go:191]: Instead of asserting the type here why not use the actual type in `apiHandler`?

File: internal/versions/rest_api_versions.go
Comment (carbonin) [inline internal/versions/rest_api_versions.go:55]: Keep the variable in the scope where it's used.

File: internal/versions/rest_api_versions.go
Comment (carbonin) [inline internal/versions/rest_api_versions.go:93]: Can we do this in sql? If not this is fine.

File: internal/versions/rest_api_versions_test.go
Comment (carbonin) [inline internal/versions/rest_api_versions_test.go:458]: I would add a test case for when someone asks for an exact version and we don't have it.

File: subsystem/agent_based_installer_client_test.go
Comment (carbonin) [reply subsystem/agent_based_installer_client_test.go:29]: I'm not even sure that needs to change.

File: deploy/wiremock/wiremock-deployment.yaml
Comment (gamli75) [reply deploy/wiremock/wiremock-deployment.yaml:18]: the org still exists but I agree that we should move it

File: internal/versions/api.go
Comment (danmanor) [reply internal/versions/api.go:191]: In second thought, I think it makes more sense to use Handler and assert, as it makes much more sense to test the api handler this way, because I can use a mock.

File: internal/versions/rest_api_versions.go
Comment (danmanor) [reply internal/versions/rest_api_versions.go:35]: Safer to validate as it goes into SQL queries don't you think ?

File: docs/hive-integration/crds/clusterImageSet.yaml
Comment (danmanor) [reply docs/hive-integration/crds/clusterImageSet.yaml:4]: OK I remember now, this is because of `agentBasedInstaller` test that uses it, which uses restAPI

File: internal/releasesources/clients.go
Comment (carbonin) [reply internal/releasesources/clients.go:36]: Looks like it's required for the wiremock stuff in subsystem, this is fine then.

File: internal/versions/api.go
Comment (carbonin) [reply internal/versions/api.go:191]: Okay, but just in case we end up ever using these APIs in kube mode maybe check that the type assertion was successful here.

File: internal/versions/api.go
Comment (carbonin) [inline internal/versions/api.go:260]: This changes the key from `releaseImage.OpenshiftVersion` to `releaseImage.Version`.
Was that intentional?

File: subsystem/authz_test.go
Comment (carbonin) [inline subsystem/authz_test.go:76]: For the future just don't make changes like this unless you happen to be making an actual change to the code. It bloats the diff and wastes time for reviewers to look through irrelevant changes.

**Review by carbonin — COMMENTED** [review]
Good work switching how we create the wiremock stubs :+1: 

Still looks like there's some work to do though between the last few comments and the failing tests.

**Comment by carbonin** [comment]
/retest

Just to see if any of these were flakes.
Some are referencing OCP versions so that's a bit suspicious IMO.

File: internal/versions/api.go
Comment (danmanor) [reply internal/versions/api.go:260]: Yes, because `openshiftVersion` may be x.y.z or x.y while `Version` is always x.y.z, there are now two different versions with the same `openshiftVersion`

File: subsystem/authz_test.go
Comment (danmanor) [reply subsystem/authz_test.go:76]: Understood, sorry

File: subsystem/operators_test.go
Comment (carbonin) [inline subsystem/operators_test.go:408]: Removing otherwise valid tests is not the correct way to deal with errors.
The error was that a release image wasn't present for 4.11.0 it feels like the correct solution would be to add the wiremock stubs to ensure a 4.11 image is present and/or update this version if necessary.

Comment (danmanor) [reply subsystem/operators_test.go:408]: I removed this test as 4.11 is `EOL` now and this test tests specific behavior of lvm in 4.11

Comment (danmanor) [reply subsystem/operators_test.go:408]: I believe this test is no longer relevant

Comment (carbonin) [reply subsystem/operators_test.go:408]: Let's keep it until we also remove the [related code](https://github.com/openshift/assisted-service/blob/7cac29ddf85d31912a091d6a9c4bd453ee6a0003/internal/operators/lvm/manifest.go#L40-L57)

Comment (danmanor) [reply subsystem/operators_test.go:408]: Yes you are right, I will change the version

Comment (danmanor) [reply subsystem/operators_test.go:408]: I just set it to 4.11, then it will take the latest 4.11 release image

File: internal/common/db.go
Comment (carbonin) [inline internal/common/db.go:542]: Can this be removed? I don't see it being used anywhere and it doesn't have tests.

File: internal/versions/rest_api_versions.go
Comment (carbonin) [reply internal/versions/rest_api_versions.go:35]: Fair enough

File: internal/common/db.go
Comment (danmanor) [reply internal/common/db.go:542]: Yes, I will remove it

**Review by carbonin — APPROVED** [review]

**Comment by danmanor** [comment]
@carbonin Thank you for your time and effort

**Pull request merged**
